polkadot/node/network/availability-recovery/src/task.rs

```diff
@@ -168,75 +227,95 @@ impl State {
 	{
 		let candidate_hash = &params.candidate_hash;
 		let already_requesting_count = requesting_chunks.len();
 
-		let mut requests = Vec::with_capacity(desired_requests_count - already_requesting_count);
+		let to_launch = desired_requests_count - already_requesting_count;
+		let mut requests = Vec::with_capacity(to_launch);
+
+		gum::trace!(
+			target: LOG_TARGET,
+			?candidate_hash,
+			"Attempting to launch {} requests",
+			to_launch
+		);
 
 		while requesting_chunks.len() < desired_requests_count {
-			if let Some(validator_index) = validators.pop_back() {
+			if let Some((chunk_index, validator_index)) = validators.pop_back() {
 				let validator = params.validator_authority_keys[validator_index.0 as usize].clone();
 				gum::trace!(
 					target: LOG_TARGET,
 					?validator,
 					?validator_index,
+					?chunk_index,
 					?candidate_hash,
 					"Requesting chunk",
 				);
 
 				// Request data.
 				let raw_request = req_res::v1::ChunkFetchingRequest {
 					candidate_hash: params.candidate_hash,
-					index: validator_index,
+					index: chunk_index,
 				};
 
 				let (req, res) = OutgoingRequest::new(Recipient::Authority(validator), raw_request);
 				requests.push(Requests::ChunkFetchingV1(req));
 
-				params.metrics.on_chunk_request_issued();
-				let timer = params.metrics.time_chunk_request();
+				params.metrics.on_chunk_request_issued(chunk_type);
+				let timer = params.metrics.time_chunk_request(chunk_type);
 
 				requesting_chunks.push(Box::pin(async move {
 					let _timer = timer;
-					match res.await {
+					let res = match res.await {
 						Ok(req_res::v1::ChunkFetchingResponse::Chunk(chunk)) =>
 							Ok(Some(chunk.recombine_into_chunk(&raw_request))),
 						Ok(req_res::v1::ChunkFetchingResponse::NoSuchChunk) => Ok(None),
-						Err(e) => Err((validator_index, e)),
-					}
+						Err(e) => Err(e),
+					};
+
+					(chunk_index, validator_index, res)
 				}));
 			} else {
 				break
 			}
 		}
 
-		sender
-			.send_message(NetworkBridgeTxMessage::SendRequests(
-				requests,
-				IfDisconnected::TryConnect,
-			))
-			.await;
+		if requests.len() != 0 {
+			sender
+				.send_message(NetworkBridgeTxMessage::SendRequests(
+					requests,
+					IfDisconnected::TryConnect,
+				))
+				.await;
+		}
 	}
 
 	/// Wait for a sufficient amount of chunks to reconstruct according to the provided `params`.
 	async fn wait_for_chunks(
 		&mut self,
+		chunk_type: &str,
 		params: &RecoveryParams,
-		validators: &mut VecDeque<ValidatorIndex>,
-		requesting_chunks: &mut FuturesUndead<
-			Result<Option<ErasureChunk>, (ValidatorIndex, RequestError)>,
-		>,
-		can_conclude: impl Fn(usize, usize, usize, &RecoveryParams, usize) -> bool,
+		retry_threshold: u32,
+		validators: &mut VecDeque<(ChunkIndex, ValidatorIndex)>,
+		requesting_chunks: &mut FuturesUndead<(
+			ChunkIndex,
+			ValidatorIndex,
+			Result<Option<ErasureChunk>, RequestError>,
+		)>,
+		can_conclude: impl Fn(usize, usize, usize, usize, usize) -> bool,
```

Comment:
I am not a fan of this, as it degrades the code readability.  Some comment would have helped.

Reply:
added comment